hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/sink/cluster/ITTestFlinkConsistentHashingClustering.java

```diff
@@ -0,0 +1,194 @@
+  private static final Map<String, String> EXPECTED_AFTER_UPSERT = new HashMap<>();
+
+  static {
+    EXPECTED_AFTER_INITIAL_INSERT.put("", "id1,,id1,Danny,23,1000,, id2,,id2,Stephen,33,2000,, "
+        + "id3,,id3,Julian,53,3000,, id4,,id4,Fabian,31,4000,, id5,,id5,Sophia,18,5000,, "
+        + "id6,,id6,Emma,20,6000,, id7,,id7,Bob,44,7000,, id8,,id8,Han,56,8000,, ]");
+    EXPECTED_AFTER_UPSERT.put("", "[id1,,id1,Danny,24,1000,, id2,,id2,Stephen,34,2000,, id3,,id3,Julian,54,3000,, "
+        + "id4,,id4,Fabian,32,4000,, id5,,id5,Sophia,18,5000,, id6,,id6,Emma,20,6000,, "
+        + "id7,,id7,Bob,44,7000,, id8,,id8,Han,56,8000,, id9,,id9,Jane,19,6000,, "
+        + "id10,,id10,Ella,38,7000,, id11,,id11,Phoebe,52,8000,,]");
+  }
+
+  @TempDir
+  File tempFile;
+
+  @Test
+  public void testScheduleSplitPlan() throws Exception {
+    TableEnvironment tableEnv = setupTableEnv();
+    prepareData(tableEnv);
+
+    Configuration conf = getDefaultConfiguration();
+    conf.setString(HoodieIndexConfig.BUCKET_INDEX_MIN_NUM_BUCKETS.key(), "4");
+    conf.setString(HoodieIndexConfig.BUCKET_INDEX_MAX_NUM_BUCKETS.key(), "8");
+    // Manually set the split threshold to trigger split in the clustering
+    conf.set(FlinkOptions.WRITE_PARQUET_MAX_FILE_SIZE, 1);
+    conf.setString(HoodieIndexConfig.BUCKET_SPLIT_THRESHOLD.key(), String.valueOf(1 / 1024.0 / 1024.0));
+    HoodieFlinkWriteClient writeClient = FlinkWriteClients.createWriteClient(conf);
+    Option<String> clusteringInstantOption = writeClient.scheduleClustering(Option.empty());
+    Assertions.assertTrue(clusteringInstantOption.isPresent());
+
+    // Validate clustering plan
+    HoodieClusteringPlan clusteringPlan = getLatestClusteringPlan(writeClient);
+    Assertions.assertEquals(4, clusteringPlan.getInputGroups().size());
+    Assertions.assertEquals(1, clusteringPlan.getInputGroups().get(0).getSlices().size());
+    Assertions.assertEquals(1, clusteringPlan.getInputGroups().get(1).getSlices().size());
+    Assertions.assertEquals(1, clusteringPlan.getInputGroups().get(2).getSlices().size());
+    Assertions.assertEquals(1, clusteringPlan.getInputGroups().get(3).getSlices().size());
+  }
+
+  @Test
+  public void testScheduleMergePlan() throws Exception {
+    TableEnvironment tableEnv = setupTableEnv();
+    prepareData(tableEnv);
+
+    Configuration conf = getDefaultConfiguration();
+    HoodieFlinkWriteClient writeClient = FlinkWriteClients.createWriteClient(conf);
+    Option<String> clusteringInstantOption = writeClient.scheduleClustering(Option.empty());
+    Assertions.assertFalse(clusteringInstantOption.isPresent());
+  }
+
+  @Test
+  public void testScheduleSortPlan() throws Exception {
+    TableEnvironment tableEnv = setupTableEnv();
+    prepareData(tableEnv);
+
+    Configuration conf = getDefaultConfiguration();
+    conf.setBoolean(FlinkOptions.BUCKET_CLUSTERING_SORT_ENABLED, true);
+    HoodieFlinkWriteClient writeClient = FlinkWriteClients.createWriteClient(conf);
```

Comment:
Not sure when to trigger the plan scheduling  in Flink pipeline.

Reply:
The existed `StreamWriteOperatorCoordinator#scheduleTableServices` already cover this.

Reply:
So the user should config the plan strategy differently then? Do they need manual config options?

Reply:
No.
If the index is a consistent bucket index, the value of `hoodie.clustering.plan.strategy.class` must be `org.apache.hudi.client.clustering.plan.strategy.FlinkConsistentBucketClusteringPlanStrategy`. `hoodie.bucket.clustering.merge.enabled` must be set to `false`.
If the index is a consistent bucket index, these two configuration values will be automatically set by the engine. The engine will also check the parameter values before submitting the job to ensure they meet the expected criteria and prevent user errors.
The only advanced configure that users might need to consider is `hoodie.bucket.clustering.sort.enabled`, which controls whether to generate regular clustering plans for buckets that are not involved in merge or split within the consistent hashing bucket index clustering plan. The default value is `false` to avoid unnecessary clustering services. If users need to, they can enable this configure.